### PR TITLE
Removing hard-coded eta dependency from PFEnergyCalibration.cc

### PR DESCRIPTION
**Pull request by spandeyehep**
This PR removes the hard-coded eta dependency from PFEnergyCalibration.cc code [which is used for PF hadron calibration].
All the hard-coded eta dependencies will be taken care of by the PFCalibration payload, making the future PF calibration updates independent of CMSSW code.
It involves following packages:
RecoParticleflow/PFClusterTools/
CondFormats/PhysicsToolsObjects/

More details can be found out in the following slides:
https://bkansal.web.cern.ch/bkansal/PFCalibration/hardcoded_dependencies_JME.pdf 

We have done the validation locally and we don't expect/see any changes at the reconstruction level from this PR.
For now, matrix level tests are likely to fail because of un-availability of compatible PFCalibration payload in the current GT.
We are in contact with AlCa/DB conveners in order to proceed with the new payload integration in the GT in parallel to this PR.

Adding @ahinzmann  , @zdemirag

### CODE REVIEW

**Comment by slava77** [comment]
> **cmsbuild** commented [an hour ago](#issuecomment-461417621)
> The code-checks are being triggered in jenkins.

@smuzaffar @mrodozov 
please check what is happening with the bot/jenkins.
It looks like the response is much slower than expected

**Comment by fabiocos** [comment]
please test

**Comment by perrotta** [comment]
What is the status of the needed GT update?
Is it included in the submitted PR 25929, by chance?

**Comment by tocheng** [comment]
> What is the status of the needed GT update?
> Is it included in the submitted PR 25929, by chance?

Hello @perrotta, the PR needs updated PF calibration to pass the test which is not in 25929.
My understanding is that the PR may needs some further update to have the internal validation succeed. AlCa is following on this, and will give suggestions soon.

**Comment by tocheng** [comment]
@spandeyehep I also mentioned this in the e-mail. 
For PFcalibration payload production, one can't set different limit for different formulas.
One can only set different limits for different variable types.

So if you want to have two kind of limits, one starting from 0 the other starting from 1,
you need to set two variable types in the ProducePFCalibrationObject (into vector), and then set two limits (into a vector), one is JetAbsEta the other is JetEt.

And when deploying the PF calibration, you need to set the correct variable type.
For e.g.,
https://github.com/cms-sw/cmssw/blob/ff23e264d4d7849c92417d6e923bb9210f1b7738/RecoParticleFlow/PFClusterTools/src/PFEnergyCalibration.cc#L508,
you should use BinningVariables::JetAbsEta

**Comment by perrotta** [comment]
@spandeyehep : is there any news here?

**Comment by spandeyehep** [comment]
Hi @perrotta ,
We had made few changes with the help of @tocheng  and also validated it.
It was being discussed on a separate email thread. 

Re-posting the last email here:

> We reverted the changes back to original. Now, we are using energy range from 0 to 1000 instead of 1 to 1000 in both the files i.e. PFEnergyCalibration.cc[1] and ProducePFCalibration.py[2]. In PFEnergyCalibration.cc code, there was already a cut on reference energy to avoid energy below 1 GeV[3].
> We re-created payload for new ranges, and validated the code and the observables using payload with updated script and old code with "103X_upgrade2018_realistic_v8" GT.
> Please find the summary plot here:
> https://bkansal.web.cern.ch/bkansal/PFCalibration/hardcoded_Eta_dependencies.pdf 
> 
> Let me know if you think new changes are okay.
> 
> Thanks and regards,
> Bhumika, Shubham
> 
> 
> Link:
> [1] https://github.com/bkansal/cmssw/blob/PFHadCalib_NoEtaDep_Nov_2018/RecoParticleFlow/PFClusterTools/src/PFEnergyCalibration.cc
> [2] https://github.com/bkansal/cmssw/blob/PFHadCalib_NoEtaDep_Nov_2018/RecoParticleFlow/PFClusterTools/test/ProducePFCalibration.py
> [3] https://github.com/bkansal/cmssw/blob/PFHadCalib_NoEtaDep_Nov_2018/RecoParticleFlow/PFClusterTools/src/PFEnergyCalibration.cc#L181
> 

The local payload is working fine with the new code and we are waiting for @tocheng to give a green light. If the new changes are okay then I'll go ahead and commit the changes in this PR.

**Comment by spandeyehep** [comment]
The new commits make the PFEnergyCalibration.cc code compatible with the payload (containing new eta dependencies).
We have re-done the validation and found no differences at RECO level. The closure can be seen here: 
https://bkansal.web.cern.ch/bkansal/PFCalibration/hardcoded_Eta_dependencies.pdf 

We have provided the corresponding payload to AlCa/DB and will be integrated in the GT.

**Comment by perrotta** [comment]
Thank you @spandeyehep 
Please notify here as soon as a pull request with the GT which includes the needed payload will become available, so that tests can be started

**Comment by srimanob** [comment]
Do I understand correctly that if we change, it will change by default, and no switch between 2 methods to make comparison, in case we need extra validation to cover all Run2.

**Comment by spandeyehep** [comment]
Hi @srimanob 
Yes that is correct, there is no switch between two methods and the changes will be made by default.

**Comment by perrotta** [comment]
please test with #26049
(Just to check whether the new GT in that PR is ok here)

**Comment by tocheng** [comment]
Looks like the failure in Matrix and AddOn tests is due to PF calibration at HLT

**Comment by fabiocos** [comment]
please test with #26049

**Comment by fabiocos** [comment]
Apparently we still have some problem to be solved. In my understanding this is a technical update, that is not going to impact in the short term the physics output of the release. So I would not keep further on hold the next pre-release for this, as it looks that more technical work is anyway needed.

**Comment by perrotta** [comment]
@spandeyehep : any update about the possible fix?

**Comment by spandeyehep** [comment]
Hi @perrotta , HLT team is working towards the creation of new/consistent payload required for HLT workflows.
We'll inform you as soon as it is done.
Adding @milosdjordjevic

**Comment by fabiocos** [comment]
@spandeyehep any news about the work for this PR?

**Comment by spandeyehep** [comment]
Hi @perrotta ,
HLT team has been facing some technical difficulties while testing/finalizing the GT.
They are working towards it. As of yet, I am not sure how long will it take to sort out the issues.
Perhaps, @milosdjordjevic might want to add some comment or give us some updates.

**Comment by spandeyehep** [comment]
Rebased and re-instantiated following codes which was removed in PR #26184 

RecoParticleFlow/PFClusterTools/test/ProducePFCalibration.py
RecoParticleFlow/PFClusterTools/test/ProducePFCalibrationObject.cc
RecoParticleFlow/PFClusterTools/test/ProducePFCalibrationObject.h
RecoParticleFlow/PFClusterTools/test/BuildFile.xml

These codes are used to generate payload for PF Charged Hadron cluster Calibration (offline & online).

**Comment by perrotta** [comment]
please test

**Comment by tocheng** [comment]
The current PF@HLT tag has wrong format. We need the tag with correct format and update GT accordingly.

**Comment by perrotta** [comment]
please abort tests
(Ok, thank you @tocheng : please let us know when that will be ready to test)

**Comment by srimanob** [comment]
@spandeyehep @tocheng 
Could you please update the status? I believe that this PR is not taking care for HLT, but also RECO, right? Then we would need it urgently.

**Comment by spandeyehep** [comment]
Hi @srimanob , the problem was with the incorrect HLT tag that was making the matrix tests fail.
HLT group has been able to sort the issue and the new tags has been queued in 106x.
Once the new GTs are available then we trigger the tests and it should work fine now.
@tocheng can comment further on this or correct me if I misunderstood something.
Thanks.

**Comment by tocheng** [comment]
> Hi @srimanob , the problem was with the incorrect HLT tag that was making the matrix tests fail.
> HLT group has been able to sort the issue and the new tags has been queued in 106x.
> Once the new GTs are available then we trigger the tests and it should work fine now.
> @tocheng can comment further on this or correct me if I misunderstood something.
> Thanks.

Hi, Lata made GT candidate.
I can't add tag to GT unless the tag is tested.
Please use GT candidate to test. Then I can add it to GT if the test pass.

**Comment by tocheng** [comment]
@spandeyehep  Lata already made a few GT candidate.
Could you please use these GT candidate to test a few failed runTheMatrix workflow.
If the failed workflows succeed using these GT candidates,
it give me confidence to put the HLT PF calibration to CMSSW release.
Thanks.

**Comment by tocheng** [comment]
@cmsbuild please test with #26415

**Comment by tocheng** [comment]
@spandeyehep @perrotta 
PR #26415 update GTs with the HLT PF calibration. I start test with the PR #26415.
Hopefully it will help this PR pass tests.

**Comment by tocheng** [comment]
@spandeyehep @fabiocos 
There is header check error
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25883/34157/headers_chks.log
Is it due to this PR?

**Comment by perrotta** [comment]
Tests can now finally run to completion, when run together with #26415. There are differences in the test outputs, but they are likely due to the modified GT: to be re-checked once the GT PR gets merged.
A few other sparse comments on the code:
- pfCalibrations and esEEInterCalib_ initialized to a null pointer and never updated: there must be a lot of unaccessed code
- There are functions with hard coded parameters: are they still accessed? If not, why did you modify some of those parameters? And since you modified them, how can you expect that output is not modified? Shouldn't they get removed once all parameters are read from GT?

**Comment by bkansal** [comment]
Thanks @tocheng

We have already done local validation for this PR and didn't get any segmentation fault while compiling the codes which are integrated, so we are not sure where the header error might be coming from.

**Comment by bkansal** [comment]
Hello @perrotta

We can't comment about esEEInterCalib_ . 
But for pfcalibrations, it is null pointer by default and it is intialized in following code -->

http://cmslxr.fnal.gov/source/RecoParticleFlow/PFProducer/plugins/PFProducer.cc#0440 

Using the command "process.particleFlowTmp.useCalibrationsFromDB = False" , functions and parameters will be taken from the constructor ( by default its true)
 
Hardcoded parameters in PFEnergyCalibration.cc code are just used for local check the changes where parameters and functions are directly taken from constructor instead payload/GT. Now we know how to directly use payload for local validation, so in future we won't need to change the hard coded parameters in constructor.

**Comment by fabiocos** [comment]
@bkansal the segmentation fault in the header check is a problem of the tool itself (@smuzaffar FYI)

**Comment by bkansal** [comment]
@fabiocos
Thanks for the information.
Please let us know, if there is anything to be done from our side.
We hope it will not cause any problem with this PR integration.

**Comment by slava77** [comment]
@spandeyehep @bkansal @tocheng 
is there a GT PR now where this PR can be tested incrementally?
The last test with #26415 showed a lot of differences; while this PR description says that ```we don't expect/see any changes at the reconstruction level from this PR.```.
I think that before we can advance with this PR there should be a testable setup that can show that there are no differences in reco.

**Comment by tocheng** [comment]
@slava77 @spandeyehep 
I have a proposal.
I can make a commit with autoCond GT including only new PF calibration at HLT.
And then @spandeyehep could include this commit to test; this will guarantee no failure and no change in reco.
Is it fine for you?

**Comment by slava77** [comment]
On 4/15/19 4:00 PM, Tongguang wrote:
> Is it fine for you?

sounds fine to me.

**Comment by tocheng** [comment]
@slava77 @spandeyehep Here is the commit
https://github.com/tocheng/cmssw/commit/7862761666506c554645ceeccbfa37028bab8a89
@spandeyehep Please add it to your PR and test. Thanks.

**Comment by spandeyehep** [comment]
Hi @tocheng , if I understood you correctly, then you want me to update autoCond.py according to tocheng@7862761 and integrate it in this PR. Is that correct?

**Comment by tocheng** [comment]
> Hi @tocheng , if I understood you correctly, then you want me to update autoCond.py according to [tocheng/cmssw@7862761](https://github.com/tocheng/cmssw/commit/7862761) and integrate it in this PR. Is that correct?

Yes. @spandeyehep Please fetch and cherry pick. Or you can even copy the file manually. It is just one file.

**Comment by spandeyehep** [comment]
Hi @tocheng , thanks for confirming.
I was facing some technical difficulties to commit the changes, which is solved now. I have updated and changed the file "autoCond.py" . But it seems like it is creating a conflict with the master branch. 
Do I have to rebase it again?

**Comment by tocheng** [comment]
@spandeyehep Don't rebase for now. The commit is only for this PR.

**Comment by spandeyehep** [comment]
@tocheng  Okay thanks a lot.

**Comment by tocheng** [comment]
@cmsbuild please test

**Comment by tocheng** [comment]
@spandeyehep Sorry, I take it back. Please solve the conflict, but make sure to use the GTs I provided in the commit. Thanks.

**Comment by spandeyehep** [comment]
Rebased for conflicting autoCond.py

**Comment by spandeyehep** [comment]
@cmsbuild please test

**Comment by christopheralanwest** [comment]
please test

@spandeyehep Only the users listed [here](http://cms-sw.github.io/categories.html) can trigger the PR tests.

**Comment by tocheng** [comment]
@spandeyehep Helllo, may bad. I forgot to update HLT GTs for you. Please include HLT GTs using this commit https://github.com/tocheng/cmssw/commit/682600f182de760339938731213c1fda6260d37a

**Comment by spandeyehep** [comment]
Hi @tocheng  I have updated the tags. Can you please check if the updated `autoCond.py` is consistent?

**Comment by tocheng** [comment]
@cmsbuild please test

**Comment by tocheng** [comment]
> Hi @tocheng I have updated the tags. Can you please check if the updated `autoCond.py` is consistent?

Yes. Thanks. Fingers crossed.

**Comment by tocheng** [comment]
@spandeyehep Please rebase your code. Thanks.

**Comment by spandeyehep** [comment]
Hi @tocheng could you please tell me which code do I have to rebase? 
Should I update autoCond.py back to original?
Thanks.

**Comment by tocheng** [comment]
> Hi @tocheng could you please tell me which code do I have to rebase?
> Should I update autoCond.py back to original?
> Thanks.

Please hold for a while. I would like to ask the green light form Slava first.

**Comment by tocheng** [comment]
@slava77 I think from the comparison of the current test, no significant difference due to moving to new HLT PF calibration. And since #26394 is just merged, GT has to contain the LHC optical function otherwise proton reconstruction will fail. The special GTs I cut for the PR don't have this LHC optical function (in order to make the test pass before #26394 merged) 
So the autoCond.py need to move to master branch if you agree.  Thanks.

**Comment by slava77** [comment]
the comparisons look promising.
However, we'll need to get a rebase/update and rerun tests anyways.

Do I understand correctly that in CMSSW_10_6_X_2019-04-17-2300 we should expect this PR (without autoCond.py changes) to run fine and show even fewer (no?) differences?

**Comment by tocheng** [comment]
> the comparisons look promising.
> However, we'll need to get a rebase/update and rerun tests anyways.
> 
> Do I understand correctly that in CMSSW_10_6_X_2019-04-17-2300 we should expect this PR (without autoCond.py changes) to run fine and show even fewer (no?) differences?

The PR can't run without autoCond changes in CMSSW_10_6_X_2019-04-17-2300.
Because the GT in the current PR don't have LHC optical function.
If they had, this PR would not pass the test before #26394 (you will get complain that two sources of LHC optical function are found: one from xml one from GT).

The GTs I made from this PR is made on the master before #26394 or before CMSSW_10_6_X_2019-04-17-2300. What I can do now is to make a new commit based the current master or CMSSW_10_6_X_2019-04-17-2300 to update data GTs to include PF calibration at HLT.

**Comment by tocheng** [comment]
> > the comparisons look promising.
> > However, we'll need to get a rebase/update and rerun tests anyways.
> > Do I understand correctly that in CMSSW_10_6_X_2019-04-17-2300 we should expect this PR (without autoCond.py changes) to run fine and show even fewer (no?) differences?
> 
> The PR can't run without autoCond changes in CMSSW_10_6_X_2019-04-17-2300.
> Because the GT in the current PR don't have LHC optical function.
> If they had, this PR would not pass the test before #26394 (you will get complain that two sources of LHC optical function are found: one from xml one from GT).
> 
> The GTs I made from this PR is made on the master before #26394 or before CMSSW_10_6_X_2019-04-17-2300. What I can do now is to make a new commit based the current master or CMSSW_10_6_X_2019-04-17-2300 to update data GTs to include PF calibration at HLT.
Then @spandeyehep  can fetch the commit and rebase the code.

I don't think fewer changes will be expected as the difference I make for GTs is PF calibration at HLT exclusively. So I expect the same relative change.

**Comment by tocheng** [comment]
> the comparisons look promising.
> However, we'll need to get a rebase/update and rerun tests anyways.
> 
> Do I understand correctly that in CMSSW_10_6_X_2019-04-17-2300 we should expect this PR (without autoCond.py changes) to run fine and show even fewer (no?) differences?

Sorry I took back what I said on the difference after rebase.
The data GTs in the current master branch already contain the PF calibration at HLT.
So when @spandeyehep rebases the PR,
he needs to make sure using the data GTs in the master branch and keep MC GTs unchanged.
This will make sure the GT change wrt master .

From this sense, there is no difference in data in comparison afte rebase.

**Comment by tocheng** [comment]
@spandeyehep 
Could you please confirm or deny that in the [current HLT PF calibration tag](https://cms-conddb.cern.ch/cmsDbBrowser/list/Prod/tags/PFCalibration_HLT_2017_25ns_Spring17_v3_hlt), you only fold the 2017 eta-dependency?
Thanks.

**Comment by tocheng** [comment]
> the comparisons look promising.
> However, we'll need to get a rebase/update and rerun tests anyways.
> 
> Do I understand correctly that in CMSSW_10_6_X_2019-04-17-2300 we should expect this PR (without autoCond.py changes) to run fine and show even fewer (no?) differences?

@slava77 Sorry I took back what I said on the difference after rebase.
The data GTs in the current master branch already contain the PF calibration at HLT,
while MC GTs in the current master branch don't.

So when @spandeyehep rebases the PR,
he needs to make sure using the data GTs in the master branch and keep MC GTs unchanged in his branch.
This will make sure the GT change wrt master is HLT PF calibration (in MC).

From this sense, there is no difference in data in comparison after rebase.
Only MC >=2015 will be affected by GT change.

**Comment by slava77** [comment]
@tocheng 
I looked a bit more carefully, and now I'm not sure I understand the differences

E.g. In the last tests, there are changes in MET uncertainties in 2016MC miniAOD.
For my previous post, I simply assumed that there are JEC or related changes
The GT changes are https://cms-conddb-prod.cern.ch/cmsDbBrowser/diff/Prod/gts/105X_mcRun2_asymptotic_v4/106X_mcRun2_asymptotic_v3
and affect only the PFCalibrationRcd label HLT.
The tests baseline was CMSSW_10_6_X_2019-04-17-1100.
Such GT change should not lead to any changes in the offline reco results.
The differences are at the numerical precision level and could be related to the changes in the code in this PR rather than in the GT.

I was previously confused why only HLT payload was mentioned.

@spandeyehep
Do I understand correctly that the cases where the hardcoded dependencies were used in the baseline were replaced with already existing functions in the baseline GT and somehow only the HLT GT did not have these functions?

Looking at the baseline GT https://cms-conddb-prod.cern.ch/cmsDbBrowser/list/Prod/gts/105X_mcRun2_asymptotic_v4
the offline payload is PFCalibrationRcd	-	PFCalibration_v10_mc.
In this PR you add
```diff
+    //added by bhumika Nov 2018
+    PFfcEta_BARRELH=3019, PFfcEta_BARRELEH = 3020,
+    PFfcEta_ENDCAPH = 3021, PFfcEta_ENDCAPEH = 3022,
+    PFfdEta_ENDCAPH = 3023, PFfdEta_ENDCAPEH =3024
```
In the payload (c85fdfbbca045de26ae6e641763c61c42ff6ab39) dump for the baseline I actually see these functions 3019 to 3024 already present .
E.g. 3024 corresponds to dEtaEndcapEH has a formula ```1*(pow((x-1.5),2))+0.6``` ;compared to the hardcoded version  ```*(0.6 + |eta-1.5|^2)```.
So, now this makes much more sense for the offline monitoring, including the small differences.

Comparing the HLT payloads in this case:
```
PFCalibration_HLT_2018_25ns_Spring19_V1 : 1b3435f6105a245e7dace748fd7fbf38286a8ad1
PFCalibration_HLT_2017_25ns_Spring17_v3_hlt : 81ab2a416bf667dff75452fd91e9d8035adf9bdb
```
I see that the first (in the baseline) has the same 20 items as the second; however 
- the last 6 (corresponding to ```//added by bhumika Nov 2018```) are not indexed (they come out with index 0), even though e.g. the last one apparently has the right functional form ```1*(pow((x-1.5),2))+0.6```. 
- oddly(?) enough, the other functions in the HLT payload changed.
So, it looks like the HLT payload was (a) incomplete for eta-dependence use and (b) changed in the primary parameterization.
Now this makes much more sense for the numerous differences in the HLT plots.

**Comment by slava77** [comment]
>  I looked a bit more carefully, and now I'm not sure I understand the differences

after such a long previous note, I should actually correct the first sentence to "and now I think I understand the differences"

**Comment by bkansal** [comment]
Dear @tocheng  @slava77 

I have checked dumping the HLT and offline PFCalibration payloads directly from GT 105X_mcRun2_asymptotic_v4 using ProducePFCalibration.py [1].

for HLT PFCalibration -->
1. Payload name -  PFCalibration_HLT_2018_25ns_Spring19_V1
2. GT - 105X_mcRun2_asymptotic_v4
3. output is https://bkansal.web.cern.ch/bkansal/PFCalibratio/2018HLTPFCalibration_payload_in105GT.txt

For offline PFCalibration -->
1. Payload name -  PFCalibration_v10_mc
2. GT - 105X_mcRun2_asymptotic_v4
3. output is https://bkansal.web.cern.ch/bkansal/PFCalibration/offlinePFCalibration_payload_in105GT.txt

But the output from HLT and offline PFCalibration are identical which is not expected (especially for the parameters of old functions). We are sure that offline PFcalibration payload is consistent with what we have provided. But we are not fully sure if our procedure for dumping payload directly from the GT is also correct or not.

@milosdjordjevic , request to Milos to please have a quick look and check if the tag included in the GT is correct to their expectations. 

[1] /afs/cern.ch/user/b/bkansal/public/codes/ProducePFCalibration.py

**Comment by tocheng** [comment]
@spandeyehep 
https://bkansal.web.cern.ch/bkansal/PFCalibratio/2018HLTPFCalibration_payload_in105GT.txt not work for me.

I don't know what is your procedure to dump the tag from DB. So I can't comment. You should have some script that can translate the condition in DB to your favored format and vice versa.
But at least you should be able to use conddb dump to translate condition into xml format.

**Comment by bkansal** [comment]
@tocheng @slava77 
Sorry for the broken link of HLT PFCalibration payload. Now I have update it and also updated the clickable link for ProducePFCalibration.py code.

**Comment by christopheralanwest** [comment]
@bkansal To make explicit the suggestion of @tocheng above, you can use the following to dump the appropriate payloads directly to xml:

```
conddb list PFCalibration_v10_mc
conddb dump -d PFCalibration_v10_mc.xml c85fdfbbca045de26ae6e641763c61c42ff6ab39
conddb list PFCalibration_HLT_2018_25ns_Spring19_V1
conddb dump -d PFCalibration_HLT_2018_25ns_Spring19_V1.xml 1b3435f6105a245e7dace748fd7fbf38286a8ad1
```

This shows that the two payloads are indeed different, and that both of your outputs above represent the `PFCalibration_v10_mc` tag.

**Comment by spandeyehep** [comment]
Hi @tocheng 
I have re-based the branch and made `autoCond.py` consistent with the master branch. Can you please confirm if it correct?
Thanks.

**Comment by bkansal** [comment]
> @bkansal To make explicit the suggestion of @tocheng above, you can use the following to dump the appropriate payloads directly to xml:
> 
> ```
> conddb list PFCalibration_v10_mc
> conddb dump -d PFCalibration_v10_mc.xml c85fdfbbca045de26ae6e641763c61c42ff6ab39
> conddb list PFCalibration_HLT_2018_25ns_Spring19_V1
> conddb dump -d PFCalibration_HLT_2018_25ns_Spring19_V1.xml 1b3435f6105a245e7dace748fd7fbf38286a8ad1
> ```
> 
> This shows that the two payloads are indeed different, and that both of your outputs above represent the `PFCalibration_v10_mc` tag.

Thanks @christopheralanwest , it helps a lot.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/25883	bdc1774 
- code changes are in line with the PR description and the follow up review
    - several files in RecoParticleFlow/PFClusterTools/test/ previously removed in #26184 are recovered and were updated appropriately for the needs of this PR changes
- jenkins tests pass and comparisons with the baseline show only small differences compatible with the numerical precision differences expected from changes in the implementation of eta-dependent functions (hardcoded vs the ones in the DB) https://github.com/cms-sw/cmssw/pull/25883#issuecomment-484470710

**Comment by tocheng** [comment]
+1

**Comment by tocheng** [comment]
@ggovi Please sign the PR when you have time. Thanks.

**Comment by fabiocos** [comment]
@ggovi this is a development needed for the UL processing, could you please check and comment or sign it? There is an update in CondFormats

**Comment by fabiocos** [comment]
please test

**Comment by ggovi** [comment]
@spandeyehep 
the addition of items in the above enum might be not backward compatible on the existing payloads in the db
Did you test that you can correctly read them?

**Comment by bkansal** [comment]
Hi @ggovi,

If we use new PFCalibration payload (i.e. PFCalibration_v10_mc) which we have already provided (and has been integrated to the current GT) then we are able to read or in other words it is compatible with the new code.
We have tested it locally, please see [1].

However, if it is old payload (e.g. PFCalibration_v9_mc in 105X_upgrade2018_realistic_v1 [2]), then our current code will not be backward compatible because of new items in enum and new functions.
This PR was opened in order to make future PFCalibrations independent of code changes.

Let me know if it answers your question.

Thanks.

[1] https://github.com/cms-sw/cmssw/pull/25883#issuecomment-467733350 
[2] https://cms-conddb.cern.ch/cmsDbBrowser/list/Prod/gts/105X_upgrade2018_realistic_v1

**Comment by ggovi** [comment]
@spandeyehep 
the required compatibility includes:
- the new code ( the one of this PR ) to be capable to read correctly ALL of the existing concerned payloads in the DB
- the old code ( any earlier release ) to be capable to read correctly the new payloads produced with this PR code
Can you please clarify if both compatibilities have been validated?

**Comment by bkansal** [comment]
Hello @ggovi 

This PR is aimed to remove hard coded part in the code and introducing new functions to be provided via payload.

>  the new code ( the one of this PR ) to be capable to read correctly ALL of the existing concerned payloads in the DB

There are new functions expected in payloads which are not present in old payloads corresponding to previous releases (e.g. PFCalibration_v9_mc in 105X_upgrade2018_realistic_v1 [2]). So yes, there is incompatibility.

>  the old code ( any earlier release ) to be capable to read correctly the new payloads produced with this PR code

Technically yes.  The old code can read the functions in new payloads which were there in previous payloads. The hardcoded part may be outdated depending on which year it is being referred to. We have tested it locally [1]. 

[1]https://github.com/cms-sw/cmssw/pull/25883#issuecomment-467733350
[2]https://cms-conddb.cern.ch/cmsDbBrowser/list/Prod/gts/105X_upgrade2018_realistic_v1

**Comment by ggovi** [comment]
@spandeyehep 
Please note that functions are not persistent members of the class, so adding functions has no impact whatsoever.
- Concerning point 1 of the compatibility requirements. Assuming that the de-serialisation works ( please clarify if you have tested this aspect ), the changes should be made in such a way that the new functions are returning some well recognisable value when called in the case of old-version payload.  In this case, the classes can be considered compatibles.
- Concerning point 2. The behaviour of the old class while reading the new payloads should be unaltered. Please clarify.

**Comment by bkansal** [comment]
Hello @ggovi,

What you say here hits our expertise and experience locally, and we would request a few clarification and help.

>> Please note that functions are not persistent members of the class...
Could you please tell us which class are you pointing to? The new functions/formulae have been added in the payload and they are read in the energyEMHad(..) function in PFEnergyCalibration.cc file in order to remove hard-coded eta dependencies.

>> Assuming that the de-serialisation works...
We are not aware of the technical term "de-serialisation" here, could you please explain it to us.
When the new code tries to look for function/formula in the "old" payload then it throws an exception, We are not sure if there are any fail-safe mechanism which can be employed in order to avoid such crash (although we are also not sure if that would be recommended by the reco group).

>> The behaviour of the old class while reading the new payloads should be unaltered. Please clarify.
Yes, to the best of our understanding. 

Thanks.
@bkansal  and @spandeyehep

File: CondFormats/PhysicsToolsObjects/interface/PerformanceResult.h [hunk 1]
```diff
@@ -31,8 +31,11 @@ class PerformanceResult {
     PFfaEta_BARRELH = 3011, PFfaEta_ENDCAPH = 3012, 
     PFfbEta_BARRELH = 3013, PFfbEta_ENDCAPH = 3014, 
     PFfaEta_BARRELEH = 3015, PFfaEta_ENDCAPEH = 3016, 
-    PFfbEta_BARRELEH = 3017, PFfbEta_ENDCAPEH = 3018 
-    
+    PFfbEta_BARRELEH = 3017, PFfbEta_ENDCAPEH = 3018,
+    //added by bhumika Nov 2018
+    PFfcEta_BARRELH=3019, PFfcEta_BARRELEH = 3020,
+    PFfcEta_ENDCAPH = 3021, PFfcEta_ENDCAPEH = 3022,
+    PFfdEta_ENDCAPH = 3023, PFfdEta_ENDCAPEH =3024
```
Comment (ggovi) [inline CondFormats/PhysicsToolsObjects/interface/PerformanceResult.h]: The relevant changes for the data persistency are the ones affecting the CondFormats. Here you have added values to the ResultType enum. Probably we don't have payloads from this class directly, but where ( in what other class depending on it ) it is used?

**Comment by bkansal** [comment]
Hi @ggovi

These enum numbers are used by PFEnergycalibration.cc[1] code.
This code passes the "ResultType" and "point" (to be evaluated from the formula) to getResult(..) function (defined in PerformancePayloadFromTFormula.cc[2]) , and getResult(..) returns the evaluated values if it finds the formula in the payload.

These enum numbers are also used to produce compatible payloads e.g. ProducePFCalibration.py[3] and ProducePFCalibrationObject.cc[4].


Thanks
Bhumika & Shubham

[1] https://github.com/spandeyehep/cmssw/blob/PFEnergyCalibration_code_eta_dependency_removal/RecoParticleFlow/PFClusterTools/src/PFEnergyCalibration.cc#L546
[2] https://cmssdt.cern.ch/lxr/source/CondFormats/PhysicsToolsObjects/src/PerformancePayloadFromTFormula.cc#0028
[3] https://github.com/spandeyehep/cmssw/blob/PFEnergyCalibration_code_eta_dependency_removal/RecoParticleFlow/PFClusterTools/test/ProducePFCalibration.py
[4] https://github.com/spandeyehep/cmssw/blob/PFEnergyCalibration_code_eta_dependency_removal/RecoParticleFlow/PFClusterTools/test/ProducePFCalibrationObject.cc#L76-L81

**Comment by ggovi** [comment]
So, can you confirm that the above enum does not appear in any data member of class used for persistent objects ? In this case, the change should not create any problem.

**Comment by bkansal** [comment]
hi @ggovi

> So, can you confirm that the above enum does not appear in any data member of class used for persistent objects ? In this case, the change should not create any problem.

To our best knowledge, these enum numbers are not used elsewhere apart from the above mentioned codes.

**Comment by ggovi** [comment]
Ok, thanks!

**Comment by ggovi** [comment]
+1

**Comment by fabiocos** [comment]
+1

**Pull request merged**